cdap-api-common/src/main/java/co/cask/cdap/api/data/format/StructuredRecord.java

```diff
@@ -71,6 +81,160 @@ public Schema getSchema() {
+  }
+
+  /**
+   * Get {@link ZonedDateTime} from field of logical type timestamp. Default {@link ZoneId} is system default.
+   *
+   * @param fieldName zoned date time field to get.
+   * @return value of the field as an object of {@link ZonedDateTime}, returns null if the field has null value
+   * logical type timestamp
+   * @throws IllegalArgumentException if the field name is not found in the map of fields or the field provided
+   *                                  is not of {@link Schema.LogicalType#TIMESTAMP_MILLIS} or
+   *                                  {@link Schema.LogicalType#TIMESTAMP_MICROS} type.
+   */
+  public ZonedDateTime getTimestamp(String fieldName) {
+    Schema.LogicalType logicalType = validateAndGetLogicalType(fieldName, schema.getField(fieldName),
+                                                               new HashSet<>(
+      Arrays.asList(Schema.LogicalType.TIMESTAMP_MILLIS.name(), Schema.LogicalType.TIMESTAMP_MICROS.name())));
+
+    if (fields.get(fieldName) == null) {
+      return null;
+    }
+
+    if (logicalType == Schema.LogicalType.TIMESTAMP_MILLIS) {
+      return getZonedDateTime(fieldName, ZoneId.systemDefault(), TimeUnit.MILLISECONDS, 1000L);
+    }
+
+    return getZonedDateTime(fieldName, ZoneId.systemDefault(), TimeUnit.MICROSECONDS, 1000000L);
+  }
+
+  /**
+   * Get {@link ZonedDateTime} from field of logical type timestamp.
+   *
+   * @param fieldName zoned date time field to get.
+   * @param zoneId    zone id for the field
+   * @return value of the field as an object of {@link ZonedDateTime}
+   * @throws IllegalArgumentException if the field name is not found in the map of fields or the field provided
+   *                                  is not of {@link Schema.LogicalType#TIMESTAMP_MILLIS} or
+   *                                  {@link Schema.LogicalType#TIMESTAMP_MICROS} type.
+   */
+  public ZonedDateTime getTimestamp(String fieldName, ZoneId zoneId) {
+    Schema.LogicalType logicalType = validateAndGetLogicalType(fieldName, schema.getField(fieldName),
+                                                               new HashSet<>(
+      Arrays.asList(Schema.LogicalType.TIMESTAMP_MILLIS.name(), Schema.LogicalType.TIMESTAMP_MICROS.name())));
+
+    if (fields.get(fieldName) == null) {
+      return null;
+    }
+
+    if (logicalType == Schema.LogicalType.TIMESTAMP_MILLIS) {
+      return getZonedDateTime(fieldName, zoneId, TimeUnit.MILLISECONDS, 1000L);
+    }
+
+    return getZonedDateTime(fieldName, zoneId, TimeUnit.MICROSECONDS, 1000000L);
+  }
+
+  /**
+   * Get zoned date and time represented by the field.
+   *
+   * @param fieldName name of the field
+   * @param zoneId    zone id for the field
+   * @return {@link ZonedDateTime} represented by field.
+   */
+  private ZonedDateTime getZonedDateTime(String fieldName, ZoneId zoneId, TimeUnit unit, long divisor) {
+    Long ts = (Long) fields.get(fieldName);
```

Comment:
Use the primitive type `long ts` instead so that the value is only unboxed once.